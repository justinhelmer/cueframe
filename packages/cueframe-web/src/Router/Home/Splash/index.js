import React from 'react';
import { MDBView, MDBMask } from 'mdbreact';
import bg from './bg.jpg';

export default () => (
  <MDBView src={bg} className="h-100">
    <MDBMask overlay="black-strong" className="flex-center flex-column text-white text-center">
      <h1 className="display-4 mx-4">Preserving memories with the power of speech.</h1>
    </MDBMask>
  </MDBView>
);