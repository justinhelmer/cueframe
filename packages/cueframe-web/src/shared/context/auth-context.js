import React, { useState, useLayoutEffect } from 'react';
import { withCookies } from 'react-cookie';

const AuthContext = React.createContext();

const AuthProvider = withCookies(({ children, cookies }) => {
  const uidCookie = cookies.get('user.id');
  const [isAuthenticated, setIsAuthenticated] = useState(uidCookie);
  const [user, setUser] = useState({});
  const [userDataReady, setUserDataReady] = useState();

  const setUserSession = async (resp) => {
    setUser(await resp.json());
    setIsAuthenticated(true);
  };

  const killUserSession = () => {
    cookies.remove('user.id');
    setUser({});
    setIsAuthenticated(false);
  }

  const login = async (email, password) => {
    const resp = await fetch('/api/user/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ email, password }),
    });

    if (resp.ok) {
      await setUserSession(resp);
    } else {
      const err = new Error('User login failed');
      err.resp = resp;
      throw err;
    }
  };

  const logout = async () => {
    const resp = await fetch('/api/user/logout', { method: 'POST' });

    if (resp.ok) {
      killUserSession();
    } else {
      const err = new Error('User logout failed');
      err.resp = resp;
      throw err;
    }
  };

  // on a page refresh, if the user has a logged in cookie, fetch user data. use layout effect so the fetch promise can be passed to context consumers
  useLayoutEffect(() => {
    if (uidCookie && !user.id) {
      const promise = fetch(`/api/user/${uidCookie}`)
        .then((resp) => {
          if (resp.ok) {
            return setUserSession(resp);
          } else {
            killUserSession();
          }
        });

      setUserDataReady(promise);
    } else {
      setUserDataReady(Promise.resolve());
    }
  }, []);

  return (
    <AuthContext.Provider value={{
      isAuthenticated,
      userDataReady,
      user,
      login,
      logout,
    }}>
      {children}
    </AuthContext.Provider>
  );
});

export { AuthContext, AuthProvider };