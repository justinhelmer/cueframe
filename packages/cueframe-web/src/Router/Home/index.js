import React, { useContext } from 'react';
import { Redirect } from 'react-router-dom';
import { AuthContext } from '../../shared/context/auth-context';
import Loadable from '../../shared/Loadable';

export default () => {
  const { isAuthenticated } = useContext(AuthContext);

  if (isAuthenticated) {
    return <Redirect to="/dashboard" />;
  }

  const Splash = Loadable(() => import('./Splash'));
  return <Splash />
}