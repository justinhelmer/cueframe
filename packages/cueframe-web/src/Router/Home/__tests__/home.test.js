import React from 'react';
import flushPromises from 'flush-promises';
import { shallow, mount } from 'enzyme';
import { MemoryRouter, Redirect } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { AuthContext } from '../../../shared/context/auth-context';
import Home from '..';

jest.useFakeTimers();

describe('<Home/>', () => {
  describe('Snapshot', () => {
    it('renders correctly for logged out users', async () => {
      const tree = renderer
        .create(
          <MemoryRouter>
            <AuthContext.Provider value={{}}>
              <Home/>
            </AuthContext.Provider>
          </MemoryRouter>
        )

      jest.advanceTimersByTime(300); // advance past delay for loadable component
      await flushPromises();
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
  
  describe('Component', () => {
    it('redirects to the dashboard for logged in users', async () => {
      // ideally we would shallow mount and dive, but enzyme doesn't currently support hooks for passing auth context.
      // therefore a shallow mount with HOCs fails to render <Redirect/>.
      // a full mount will render the redirect, but also actually fire it (since not shallow).
      // that's ok since we are using the memory router, but a warning will be displayed. Suppress it.
      jest.spyOn(console, 'error').mockImplementation(() => {}); // noop

      const wrapper = mount(
        <MemoryRouter>
          <AuthContext.Provider value={{ isAuthenticated: true }}>
            <Home/>
          </AuthContext.Provider>
        </MemoryRouter>
      );

      expect(wrapper.find(Redirect)).toHaveProp('to', '/dashboard');
      console.error.mockRestore();
    });
  });
});