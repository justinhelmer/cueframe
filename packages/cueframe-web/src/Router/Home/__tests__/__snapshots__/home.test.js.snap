// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home/> Snapshot renders correctly for logged out users 1`] = `
<div
  className="view h-100"
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundImage": "url(\\"bg.jpg\\")",
      "backgroundPosition": "center center",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "cover",
      "height": "100vh",
    }
  }
>
  <div
    className="mask rgba-black-strong flex-center flex-column text-white text-center"
  >
    <h1
      className="display-4 mx-4"
    >
      Preserving memories with the power of speech.
    </h1>
  </div>
</div>
`;
